src/DripsDeployer.sol

```diff
@@ -0,0 +1,412 @@
+    bytes32[] internal _moduleSalts;
+    address public immutable initialOwner;
+
+    function args() public view returns (bytes memory) {
+        return abi.encode(initialOwner);
+    }
+
+    constructor(address initialOwner_) {
+        // slither-disable-next-line missing-zero-check
+        initialOwner = initialOwner_;
+        _transferOwnership(initialOwner);
+    }
+
+    function deployModules(
+        Module[] calldata modules1,
+        Module[] calldata modules2,
+        Module[] calldata modules3,
+        Module[] calldata modules4
+    ) public onlyOwner {
+        _deployModules(modules1);
+        _deployModules(modules2);
+        _deployModules(modules3);
+        _deployModules(modules4);
+    }
+
+    function _deployModules(Module[] calldata modules) internal {
+        for (uint256 i = 0; i < modules.length; i++) {
+            Module calldata module = modules[i];
+            _moduleSalts.push(module.salt);
+            // slither-disable-next-line reentrancy-eth,reentrancy-no-eth
+            Create3.deploy(module.amount, module.salt, module.initCode);
+        }
+    }
+
+    function moduleSalts() public view returns (bytes32[] memory) {
+        return _moduleSalts;
+    }
+
+    function moduleAddress(bytes32 salt) public view returns (address addr) {
+        return Create3.computeAddress(salt);
+    }
+}
+
+abstract contract BaseModule {
+    address public immutable dripsDeployer;
+    bytes32 public immutable moduleSalt;
+
+    constructor(address dripsDeployer_, bytes32 moduleSalt_) {
+        require(
+            address(this) == Create3.computeAddress(moduleSalt_, dripsDeployer_),
+            "Invalid module deployment salt"
+        );
+        dripsDeployer = dripsDeployer_;
+        moduleSalt = moduleSalt_;
+    }
+
+    function args() public view virtual returns (bytes memory);
+
+    function _moduleAddress(bytes32 salt) internal view returns (address addr) {
+        return Create3.computeAddress(salt, dripsDeployer);
+    }
+
+    modifier onlyModule(bytes32 salt) {
+        require(msg.sender == _moduleAddress(bytes32(salt)));
+        _;
+    }
+}
+
+abstract contract ContractDeployerModule is BaseModule {
+    bytes32 public immutable salt = "deployment";
+
+    function deployment() public view returns (address) {
+        return Create3.computeAddress(salt);
+    }
+
+    function deploymentArgs() public view virtual returns (bytes memory);
+
+    function _deployContract(bytes memory creationCode) internal {
+        Create3.deploy(0, salt, abi.encodePacked(creationCode, deploymentArgs()));
+    }
+}
+
+abstract contract ProxyDeployerModule is BaseModule {
+    bytes32 public immutable proxySalt = "proxy";
+    address public proxyAdmin;
+    address public logic;
+
+    function proxy() public view returns (address) {
+        return Create3.computeAddress(proxySalt);
+    }
+
+    function proxyArgs() public view returns (bytes memory) {
+        return abi.encode(logic, proxyAdmin);
```

Comment:
Maybe adding a sanity check here?
```solidity
require(logic != address(0) && proxyAdmin != address(0));
```

Because it requires the inherited contract to first set this variable before `_deployProxy` can be called.

Reply:
`_deployProxy` sets those variables, so they can't be not set. The worst that could happen is not calling `_deployProxy`, but that's tricky to catch without explicity calling `_verifyDeployProxyCalled` or something similar in the module. It'd be optimal if `_deployProxy` was just a constructor, because the constructor would require it to be called, but then it'd be difficult to do any preparation before deploying the proxy. Adding the safety check would only generate a nicer error in the script when it called `proxyArgs` on a badly written module. We probably should add safety checks for `logic()` and `proxyAdmin()` then too. I feel that it's not worth the effort, how about you?